### ISSUE TITLE
support for Optifine 

### ISSUE THREAD

**Issue by jugendhausldbg** [mindstorm38/portablemc#20]
Hi there,
is it possible to add support for Optifine, or is there already a way to enable Optifine?
thanks for your amazing work

**Comment by mindstorm38** [comment]
Hello! Thank you for supporting this project!

Optifine is not currently supported through any addon, however this is a great idea of add-on.

**Comment by GoodDay360** [comment]
@jugendhausldbg You can install Optifine with Fabric.

**Comment by mindstorm38** [comment]
Yes, optifine with fabric is a good idea. If you really want optifine only you can use the optifine installer and then use the version name (something like `1.16.2-OptiFine_HD_U_G3`) to start the version with `start 1.16.2-OptiFine_HD_U_G3`. This should work.

**Comment by mindstorm38** [comment]
I contacted sp614x (OptiFine's author) to ask him for permission to make an addon like this.

He's ok if and only if OptiFine is not bundled nor distributed with PortableMC's package. Hopefully, he seems ok with the idea of an add-on that will download the mod from the website.

This message is just a note for the future, the add-on is not in development for now.

**Comment by GoodDay360** [comment]
Btw If you need a way to download Optifine and list all of Optifine versions. I could provide you with the API.

**Comment by mindstorm38** [comment]
Yes why not it would be cool because my current idea is to parse the web page, which is not really easy and lightweight.

**Comment by GoodDay360** [comment]
@mindstorm38 Check https://github.com/GoodDay360/Optifine-API.

**Comment by mindstorm38** [comment]
Who is maintaining the API?

**Comment by GoodDay360** [comment]
@mindstorm38 https://rapidapi.com/user/nitroxenon. The website is broken I used backdoor requests to subscribe to the API.

**Comment by WoMspace** [comment]
hello! I'm a shader developer, and a moderater over on the shaderLABS discord.
Downloading Optifine programatically from the website is perfectly fine, and sp614x will have no issues with that.

As for an API, major releases are few and far between, so hosting and maintaining a list of URLs somewhere online would be trivial. just `curl <someurl>/optifine_links | grep <version>` or something when an install is requested.
You might want to take a look at [vivecraft](https://github.com/jrbudda/Vivecraft_118), and more specifically [install.py](https://github.com/jrbudda/Vivecraft_118/blob/master/install.py#L15) and [minecriftversion.py](https://github.com/jrbudda/Vivecraft_118/blob/master/minecriftversion.py).
Alternatively, would it be so hard to ask users to download it and just provide the jar file?

Regardless, if you would like any assistance regarding optifine (including downloading) head over to the [shaderLABS discord server](https://discord.gg/RpzWN9S)

**Comment by mindstorm38** [comment]
Thank you for support! I'll definitely look into it in the future

**Comment by GoodDay360** [comment]
I wonder if PMC supports optifine yet, because it seems like this request is still open. Anyway ago I provided the api for getting optifine but it is not well maintained.
However a few days ago I managed to web scrape the optifine site and get all version info. 
Sauce: https://github.com/GoodDay360/OptiPy

Optifine author seems to be fine about it and he also gives me a direct link for downloading the file.

**Comment by mindstorm38** [comment]
Optifine is not yet supported, and I still wonder if this is a good idea, mainly because Optifine in itself can be installed really quickly as a forge mod, and forge can be installed really easily now... This is not definitive, and I'm still opened to arguments, because I'm afraid it will require some hard work.

**Comment by GoodDay360** [comment]
> Optifine is not yet supported, and I still wonder if this is a good idea, mainly because Optifine in itself can be installed really quickly as a forge mod, and forge can be installed really easily now... This is not definitive, and I'm still opened to arguments, because I'm afraid it will require some hard work.

It doesn't require much work to implement this. Just download and write optifine into the user mod folder if they run it with forge especially the version that fabric doesn't support if they want to. This saves so much time in some cases instead of opening the browser every time u want optifine on a different version. Caches downloaded optifine if possible if the user doesn't have internet, this may help in case users don't save their own. (Caches doesn't mean bundle the file so it is probably safe to do it).

More than that you can also make an optifine standalone installation too if u have some free time. Launch optifine directly without forge. This gonna take a while to wrap everything.

**Comment by mindstorm38** [comment]
For now, I'll stay away from the mess of managing the "mods" directory, because this would require checking for already installed version, disabling old one, all of this is not trivial at all because mods are not uniquely defined by the file name.

So it seems way more realistic to me to support standalone installation, even if this is still complicated in itself.

**Comment by pi-dev500** [comment]
I wrote a complete install script for optifine which works together with portablemc. If you are interested (the script is currently tested with more than 20 different version, but as I coded it very recently, I can't say that it is stable). If you are interested , I can post it.

**Comment by mindstorm38** [comment]
Wow, insane! I'm really interested in seeing that if you agree.

**Comment by pi-dev500** [comment]
It's a dirty python rewrite of the decompiled optifine installer with additionnal features, using the API of portablemc. I can send it today, but I'll try to cleanup it before sending it in the week.

**Comment by gabdevele** [comment]
Hi @theorzr, are there plans to add the ability to launch Optifine directly with any version?
For example:
`portablemc start 1.8.9 --optifine`

Thanks, and congratulations on this project.
